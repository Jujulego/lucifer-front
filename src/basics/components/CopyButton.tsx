import React, { MouseEvent, useEffect, useState } from 'react';
import copy from 'copy-to-clipboard';

import {
  IconButton, Tooltip,
  IconButtonProps
} from '@material-ui/core';
import FileCopyIcon from '@material-ui/icons/FileCopy';

// Types
export interface CopyButtonProps extends IconButtonProps {
  text: string;
  format?: string;

  tooltip?: string;
  tooltipTimeout?: number;

  onCopied?: () => void;
}

// Component
const CopyButton = (props: CopyButtonProps) => {
  // Props
  const {
    text, format = "text/plain",
    tooltip = "Copi√©",
    tooltipTimeout = 1500,
    onCopied, onClick,
    ...btn
  } = props;

  // State
  const [showTooltip, setShowTooltip] = useState(false);

  // Handler
  const handleCopy = (event: MouseEvent<HTMLButtonElement>) => {
    if (onClick) onClick(event);

    // Copy
    const success = copy(text, { format });

    if (success) {
      setShowTooltip(true);
      if (onCopied) onCopied();
    }
  };

  // Effect
  useEffect(() => {
    if (showTooltip) {
      const id = setTimeout(() => setShowTooltip(false), tooltipTimeout);
      return () => clearTimeout(id);
    }
  }, [showTooltip, tooltipTimeout]);

  // Render
  return (
    <Tooltip
      title={tooltip} open={showTooltip}
      onOpen={() => setShowTooltip(true)}
      onClose={() => setShowTooltip(false)}
      disableFocusListener disableHoverListener disableTouchListener
    >
      <IconButton {...btn} onClick={handleCopy}>
        <FileCopyIcon fontSize="small" />
      </IconButton>
    </Tooltip>
  );
};

export default CopyButton;