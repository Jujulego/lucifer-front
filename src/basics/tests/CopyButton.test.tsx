import React from 'react';
import copy from 'copy-to-clipboard';
import { createShallow, createMount } from '@material-ui/core/test-utils';

import CopyButton from '../components/CopyButton';

// Mocks
jest.mock('copy-to-clipboard');

// Setup
let mount: ReturnType<typeof createMount>;
let shallow: ReturnType<typeof createShallow>;

beforeAll(() => {
  mount = createMount();
  shallow = createShallow();
});

afterAll(() => {
  mount.cleanUp();
  jest.restoreAllMocks();
});

beforeEach(() => {
  (copy as jest.Mock).mockReset();
});

// Tests
it('should render correctly', () => {
  // Render
  const wrapper = shallow(
    <CopyButton text={'test'} />
  );

  // Check elements
  expect(wrapper).toMatchSnapshot();
});

it('should copy on click', () => {
  const spy = jest.fn();
  (copy as jest.Mock).mockImplementation(() => true);

  // Render
  const wrapper = mount(
    <CopyButton
      text={'test'}
      onCopied={spy}
    />
  );

  // Get button
  const button = wrapper.find('button');
  expect(button).toHaveLength(1);

  // Test event
  button.simulate('click');

  expect(copy).toHaveBeenCalledTimes(1);
  expect(copy).toHaveBeenCalledWith('test', { format: 'text/plain' });

  expect(spy).toHaveBeenCalledTimes(1);
})
