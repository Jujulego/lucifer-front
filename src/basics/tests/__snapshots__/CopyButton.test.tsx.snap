// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<WithStyles(ForwardRef(Tooltip))
  disableFocusListener={true}
  disableHoverListener={true}
  disableTouchListener={true}
  onClose={[Function]}
  onOpen={[Function]}
  open={false}
  title="CopiÃ© !"
>
  <WithStyles(ForwardRef(IconButton))
    onClick={[Function]}
  >
    <FileCopyIcon
      fontSize="small"
    />
  </WithStyles(ForwardRef(IconButton))>
</WithStyles(ForwardRef(Tooltip))>
`;
